packages/api/apps/processes.mts

```diff
@@ -62,23 +63,84 @@ export function deleteAppProcess(appId: string, process: ProcessType) {
   processes.del(appId, process);
 }
 
+export async function waitForProcessToComplete(process: AppProcessType): Promise<void> {
+  if (process.process.exitCode !== null) {
+    return;
+  }
+
+  return new Promise((resolve, reject) => {
+    process.process.once("exit", () => {
+      resolve();
+    });
+    process.process.once("error", (err) => {
+      reject(err);
+    });
+  })
+}
+
 /**
  * Runs npm install for the given app.
  *
  * If there's already a process running npm install, it will return that process.
  */
 export function npmInstall(
   appId: string,
-  options: Omit<Parameters<typeof execNpmInstall>[0], 'cwd'>,
+  options: Omit<Partial<Parameters<typeof execNpmInstall>[0]>, 'cwd'> & { onStart?: () => void },
 ) {
-  if (!processes.has(appId, 'npm:install')) {
-    processes.set(appId, {
-      type: 'npm:install',
-      process: execNpmInstall({ cwd: pathToApp(appId), ...options }),
-    });
+  const runningProcess = processes.get(appId, 'npm:install');
+  if (runningProcess) {
+    return runningProcess;
+  }
+
+  wss.broadcast(`app:${appId}`, 'deps:install:status', { status: 'installing' });
+  if (options.onStart) {
+    options.onStart();
   }
 
-  return processes.get(appId, 'npm:install');
+  const newlyStartedProcess: NpmInstallProcessType = {
+    type: 'npm:install',
+    process: execNpmInstall({
+      ...options,
+
+      cwd: pathToApp(appId),
+      stdout: (data) => {
+        wss.broadcast(`app:${appId}`, 'deps:install:log', {
+          log: { type: 'stdout', data: data.toString('utf8') },
+        });
+
+        if (options.stdout) {
+          options.stdout(data);
+        }
+      },
+      stderr: (data) => {
+        wss.broadcast(`app:${appId}`, 'deps:install:log', {
+          log: { type: 'stderr', data: data.toString('utf8') },
+        });
+
+        if (options.stderr) {
+          options.stderr(data);
+        }
+      },
+      onExit: (code) => {
+        // We must clean up this process so that we can run npm install again
+        deleteAppProcess(appId, 'npm:install');
+
+        wss.broadcast(`app:${appId}`, 'deps:install:status', {
+          status: code === 0 ? 'complete' : 'failed',
+          code,
+        });
+
+        if (code === 0) {
+          wss.broadcast(`app:${appId}`, 'deps:status:response', {
+            nodeModulesExists: true,
+          });
+        }
+      },
+    }),
+  };
+  processes.set(appId, newlyStartedProcess);
+
+  return newlyStartedProcess;
 }
```

Comment:
This might overload this process abstraction a bit, but I moved all the socket message emitting logic here - this is what makes sure these always get emitted for all cases where `npm install` is run.

I could potentially make another level of abstraction here if desired, though I don't think in practice there would be a reason one would want to call `npmInstall` without the socket logic associated.